alpaca/Dockerfile

```diff
@@ -1,6 +1,20 @@
 # syntax=docker/dockerfile:experimental
+FROM local/java:latest as alpacabuild
+
+ENV ALPACA_REPO="https://github.com/jhu-idc/Alpaca"
+ENV ALPACA_BRANCH="Alpaca-1.0.3-jhu"
+
+RUN mkdir /build && \
+    cd /build && \
+    git clone -b ${ALPACA_BRANCH} ${ALPACA_REPO} . && \
+    cd /build && \
+    ./gradlew install
+
 FROM local/karaf:latest
 
+# Copy locally built Alpaca artifacts
+COPY --from=alpacabuild /root/.m2/repository /root/.m2/repository
```

Comment:
This results in adding a couple dozen MB to the resulting image, as the Karaf installation process _copies_ maven artifacts into Karaf's internal bundles directory.  The end result is that build-related jars are still int he published image's `.m2` directory, and copies of the resulting Alpaca jars are both in `.m2` and `/opt/karaf`

An optimization would be to _mount_ `.m2` from the alpacabuild container instead of `COPY`.  See the pattern in the [houdini build](https://github.com/jhu-idc/idc-isle-buildkit/blob/main/houdini/Dockerfile#L9), the solution would essentially be something like:

```
FROM local/java:latest as alpacabuild

// do build stuff

FROM local/karaf:latest

RUN --mount=type=bind,from=alpacabuild,source=/root/,m2/repository,target=/root/m2/repository \
  /bin/start... // the rest of the RUN statement
 ```

Reply:
I get a weird error when adding `--mount=type=bind,from=alpacabuild,source=/root/.m2/repository,target=/root/.m2/repository` to the RUN statement:

The output below is emitted by the following:
```
RUN --mount=type=bind,from=alpacabuild,source=/root/.m2/repository,target=/root/.m2/repository \
#RUN \
    find /root/.m2/repository -type f && \
    bin/start && \
    ALPACA_VERSION=1.0.3 && \
    ACTIVEMQ_VERSION=5.15.0 && \
    CAMEL_VERSION=2.20.4 && \
    bin/client -r 10 -d 5  "feature:repo-add mvn:ca.islandora.alpaca/islandora-karaf/${ALPACA_VERSION}/xml/features" && \
    bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.activemq/activemq-karaf/${ACTIVEMQ_VERSION}/xml/features" && \
    bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.camel.karaf/apache-camel/${CAMEL_VERSION}/xml/features" && \
    bin/client -r 10 -d 5  "feature:install fcrepo-service-activemq" && \
    bin/client -r 10 -d 5  "feature:install fcrepo-service-camel" && \
    bin/client -r 10 -d 5  "feature:install islandora-http-client" && \
    bin/stop && \
    rm -rf instances/*
```

```
#12 [stage-1 3/8] RUN --mount=type=bind,from=alpacabuild,source=/root/.m2/re...
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-connector-derivative/maven-metadata-local.xml
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-connector-derivative/1.0.3/islandora-connector-derivative-1.0.3-configuration.cfg
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-connector-derivative/1.0.3/islandora-connector-derivative-1.0.3.jar
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-connector-derivative/1.0.3/islandora-connector-derivative-1.0.3-sources.jar
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-connector-derivative/1.0.3/islandora-connector-derivative-1.0.3.pom
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-connector-derivative/1.0.3/islandora-connector-derivative-1.0.3-javadoc.jar
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-http-client/maven-metadata-local.xml
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-http-client/1.0.3/islandora-http-client-1.0.3-sources.jar
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-http-client/1.0.3/islandora-http-client-1.0.3.pom
#12 0.277 /root/.m2/repository/ca/islandora/alpaca/islandora-http-client/1.0.3/islandora-http-client-1.0.3-configuration.cfg
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-http-client/1.0.3/islandora-http-client-1.0.3-javadoc.jar
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-http-client/1.0.3/islandora-http-client-1.0.3.jar
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-event-support/maven-metadata-local.xml
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-event-support/1.0.3/islandora-event-support-1.0.3-sources.jar
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-event-support/1.0.3/islandora-event-support-1.0.3-javadoc.jar
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-event-support/1.0.3/islandora-event-support-1.0.3.jar
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-event-support/1.0.3/islandora-event-support-1.0.3.pom
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-fcrepo/maven-metadata-local.xml
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-fcrepo/1.0.3/islandora-indexing-fcrepo-1.0.3.jar
#12 0.278 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-fcrepo/1.0.3/islandora-indexing-fcrepo-1.0.3-sources.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-fcrepo/1.0.3/islandora-indexing-fcrepo-1.0.3-javadoc.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-fcrepo/1.0.3/islandora-indexing-fcrepo-1.0.3-configuration.cfg
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-fcrepo/1.0.3/islandora-indexing-fcrepo-1.0.3.pom
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-karaf/maven-metadata-local.xml
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-karaf/1.0.3/islandora-karaf-1.0.3-features.xml
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-karaf/1.0.3/islandora-karaf-1.0.3.pom
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-karaf/1.0.3/islandora-karaf-1.0.3.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-karaf/1.0.3/islandora-karaf-1.0.3-sources.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-karaf/1.0.3/islandora-karaf-1.0.3-javadoc.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-triplestore/maven-metadata-local.xml
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-triplestore/1.0.3/islandora-indexing-triplestore-1.0.3-sources.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-triplestore/1.0.3/islandora-indexing-triplestore-1.0.3.pom
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-triplestore/1.0.3/islandora-indexing-triplestore-1.0.3-configuration.cfg
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-triplestore/1.0.3/islandora-indexing-triplestore-1.0.3.jar
#12 0.279 /root/.m2/repository/ca/islandora/alpaca/islandora-indexing-triplestore/1.0.3/islandora-indexing-triplestore-1.0.3-javadoc.jar
#12 5.982 retrying (attempt 1) ...
#12 7.879 Adding feature url mvn:ca.islandora.alpaca/islandora-karaf/1.0.3/xml/features
#12 8.902 Adding feature url mvn:org.apache.activemq/activemq-karaf/5.15.0/xml/features
#12 8.945 Error executing command: Error resolving artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0: [Could not transfer artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0 from/to central (https://repo1.maven.org/maven2/): /root/.m2/repository/org/apache/activemq/activemq-karaf/5.15.0/activemq-karaf-5.15.0-features.xml.part.lock (No such file or directory), Could not transfer artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0 from/to spring.ebr.release (http://repository.springsource.com/maven/bundles/release/): /root/.m2/repository/org/apache/activemq/activemq-karaf/5.15.0/activemq-karaf-5.15.0-features.xml.part.lock (No such file or directory), Could not transfer artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0 from/to spring.ebr.external (http://repository.springsource.com/maven/bundles/external/): /root/.m2/repository/org/apache/activemq/activemq-karaf/5.15.0/activemq-karaf-5.15.0-features.xml.part.lock (No such file or directory), Could not transfer artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0 from/to gemini (http://zodiac.springsource.com/maven/bundles/release/): /root/.m2/repository/org/apache/activemq/activemq-karaf/5.15.0/activemq-karaf-5.15.0-features.xml.part.lock (No such file or directory)] : mvn:org.apache.activemq/activemq-karaf/5.15.0/xml/features
#12 ERROR: executor failed running [/bin/sh -c find /root/.m2/repository -type f &&     bin/start &&     ALPACA_VERSION=1.0.3 &&     ACTIVEMQ_VERSION=5.15.0 &&     CAMEL_VERSION=2.20.4 &&     bin/client -r 10 -d 5  "feature:repo-add mvn:ca.islandora.alpaca/islandora-karaf/${ALPACA_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.activemq/activemq-karaf/${ACTIVEMQ_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.camel.karaf/apache-camel/${CAMEL_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:install fcrepo-service-activemq" &&     bin/client -r 10 -d 5  "feature:install fcrepo-service-camel" &&     bin/client -r 10 -d 5  "feature:install islandora-http-client" &&     bin/stop &&     rm -rf instances/*]: runc did not terminate sucessfully
------
 > [stage-1 3/8] RUN --mount=type=bind,from=alpacabuild,source=/root/.m2/repository,target=/root/.m2/repository     find /root/.m2/repository -type f &&     bin/start &&     ALPACA_VERSION=1.0.3 &&     ACTIVEMQ_VERSION=5.15.0 &&     CAMEL_VERSION=2.20.4 &&     bin/client -r 10 -d 5  "feature:repo-add mvn:ca.islandora.alpaca/islandora-karaf/${ALPACA_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.activemq/activemq-karaf/${ACTIVEMQ_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.camel.karaf/apache-camel/${CAMEL_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:install fcrepo-service-activemq" &&     bin/client -r 10 -d 5  "feature:install fcrepo-service-camel" &&     bin/client -r 10 -d 5  "feature:install islandora-http-client" &&     bin/stop &&     rm -rf instances/*:
------
failed to solve with frontend dockerfile.v0: failed to solve with frontend gateway.v0: rpc error: code = Unknown desc = failed to build LLB: executor failed running [/bin/sh -c find /root/.m2/repository -type f &&     bin/start &&     ALPACA_VERSION=1.0.3 &&     ACTIVEMQ_VERSION=5.15.0 &&     CAMEL_VERSION=2.20.4 &&     bin/client -r 10 -d 5  "feature:repo-add mvn:ca.islandora.alpaca/islandora-karaf/${ALPACA_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.activemq/activemq-karaf/${ACTIVEMQ_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:repo-add mvn:org.apache.camel.karaf/apache-camel/${CAMEL_VERSION}/xml/features" &&     bin/client -r 10 -d 5  "feature:install fcrepo-service-activemq" &&     bin/client -r 10 -d 5  "feature:install fcrepo-service-camel" &&     bin/client -r 10 -d 5  "feature:install islandora-http-client" &&     bin/stop &&     rm -rf instances/*]: runc did not terminate sucessfully

> Task :alpaca:build FAILED

FAILURE: Build failed with an exception.
```

The output of the `find` command is the same regardless of whether a `COPY` or `RUN` a run bind mount is used.

Reply:
Using the options to the `RUN` command produces the weird error when installing the artifacts:

```
Error resolving artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0: [Could not transfer artifact org.apache.activemq:activemq-karaf:xml:features:5.15.0 from/to central (https://repo1.maven.org/maven2/): /root/.m2/repository/org/apache/activemq/activemq-karaf/5.15.0/activemq-karaf-5.15.0-features.xml.part.lock (No such file or directory)
```

Reply:
So I think for now I'm going to leave the `COPY` as-is.  The issue with the `RUN` bind mount approach is not readily understandable, so I'm not sure what the fix would be.

Reply:
@birkland ^^

Reply:
I'm wondering if it's read-only, yet the karaf install process somehow wants to write stuff to `.m2`.  Anyway, I agree.  Not worth time to deal with this now.

Reply:
@birkland Oh, that's a possibility.  I didn't consider that Karaf would want to write to .m2; that'd be weird, but it _is_ Karaf.